src/dataplane-protocol/src/record.rs

```diff
@@ -20,67 +19,76 @@ use crate::core::Version;
 use crate::derive::Decode;
 use crate::derive::Encode;
 
-use crate::batch::DefaultBatch;
+use crate::batch::Batch;
 use crate::Offset;
 
-pub type DefaultRecord = Record<DefaultAsyncBuffer>;
-
 /// maximum text to display
 static MAX_STRING_DISPLAY: Lazy<usize> = Lazy::new(|| {
     let var_value = std::env::var("FLV_MAX_STRING_DISPLAY").unwrap_or_default();
     var_value.parse().unwrap_or(16384)
 });
 
-/// slice that can works in Async Context
-pub trait AsyncBuffer {
-    fn len(&self) -> usize;
+/// A key for determining which partition a record should be sent to.
+///
+/// This type is used to support conversions from any other type that
+/// may be converted to a `Vec<u8>`, while still allowing the ability
+/// to explicitly state that a record may have no key (`RecordKey::NULL`).
+///
+/// # Examples
+///
+/// ```
+/// # use fluvio_dataplane_protocol::record::RecordKey;
+/// let key = RecordKey::NULL;
+/// let key: RecordKey = "Hello, world!".into();
+/// let key: RecordKey = String::from("Hello, world!").into();
+/// let key: RecordKey = vec![1, 2, 3, 4].into();
+/// ```
+pub struct RecordKey(RecordKeyInner);
+
+impl RecordKey {
+    pub const NULL: Self = Self(RecordKeyInner::Null);
+
+    fn into_option(self) -> Option<RecordData> {
+        match self.0 {
+            RecordKeyInner::Key(key) => Some(key),
+            RecordKeyInner::Null => None,
+        }
+    }
 }
 
-pub trait Records {}
-
-#[derive(Default)]
-pub struct DefaultAsyncBuffer(Bytes);
+enum RecordKeyInner {
+    Null,
+    Key(RecordData),
+}
 
-impl DefaultAsyncBuffer {
```

Comment:
need to preserve `DefaultAsyncBuffer` for compatibility.  Mark as Deprecated

Reply:
I believe that due to other changes in this PR that we will need a breaking change either way. Are you referring to crate compatibility or protocol compatibility? Because for either of those cases:

- Crate compatibility is already being broken, so it would not hurt to remove this type, and
- Protocol compatibility should be maintained even though we have different type names (i.e. the Encode and Decode implementations are the same).

The only reason I'm pushing back on this is that I have not only renamed this but I've also changed a bunch of the conversions and done other breaking things to this type, and the only good way to "preserve compatibility" would be to undo all of those changes and have that code hang out in a module somewhere where it's probably never going to be used. I think it is a better choice to just lean into the breaking change in this case.

Reply:
Here's a summary of changes after latest discussion:

- Renamed DefaultAsyncBuffer to RecordData
- Added a type alias from DefaultAsyncBuffer to RecordData so that any crates importing DefaultAsyncBuffer will still work
  - Added a #[deprecated]` tag to the DefaultAsyncBuffer type alias so those users get warnings.